firedrake/preconditioners/pmg.py

```diff
@@ -62,12 +62,38 @@ def coarsen_element(ele):
         degree = ele.degree()
         family = ele.family()
 
-        if family == "Discontinuous Galerkin" and degree == 0:
+        N = degree
+        try:
+            N, = set(N)
+        except TypeError:
+            pass
+        except ValueError:
+            raise NotImplementedError("Different degrees on TensorProductElement")
+
+        if family == "Discontinuous Galerkin" and N == 0:
             raise ValueError
-        elif degree == 1:
+        elif N == 1:
             raise ValueError
 
-        return ele.reconstruct(degree=degree // 2)
+        return PMGBase.reconstruct_degree(ele, N // 2)
+
+    @staticmethod
+    def reconstruct_degree(ele, N):
+        """
+        Reconstruct a given element, modifying its polynomial degree.
+
+        May only set the same degree on VectorElements and TensorProductElements.
+
+        :arg ele: a :class:`ufl.FiniteElement` to reconstruct.
+        :arg N: an integer degree.
+        """
+        if isinstance(ele, (VectorElement, TensorElement)):
+            sub = ele.sub_elements()
+            return VectorElement(PMGBase.reconstruct_degree(sub[0], N), dim=len(sub))
+        elif isinstance(ele, TensorProductElement):
+            return TensorProductElement(*(PMGBase.reconstruct_degree(sub, N) for sub in ele.sub_elements()))
+        else:
+            return ele.reconstruct(degree=N)
 
```

Comment:
This work is interesting, and I've needed similar things for P1PC-type things in tensor product cells.  It seems like the reconstruction should/could be done at the UFL level instead?

Reply:
So `P1PC` is a subclass of `PMGPC` but with `coarsen_element` overridden. One solution is to get @pbrubeck to modify that the overridden `coarsen_element` appropriately in this PR or a later PR also. But I think what @rckirby is asking is if this knowledge should go back into UFL. @wence- , any thoughts?

Reply:
I can easily extend this coarsening to `P1PC`. But a question arises: do we want the lowest order for a familty (e.g. degree = 0 in the case of DG) or stick to the naming and have always degree = 1?

Reply:
I like Rob's idea of defaulting to 1 and having it controllable via PETSc options.

Reply:
> Unclear. DG(0) is lowest order, but some DG methods require degree 1 to actually work. Thought: default to 1, but is there a way to have a PETSc option specifying the degree you want to reduce to so it could be overridden? Similar use case -- maybe you've got a good preconditioner for your degree 2 case and want to use it instead of degree 1.

I like your idea @rckirby. One way to do this is changing `coarsen_element` to a class method and having the coarse degree as an attribute set during the initialization. I don't think there would be any negative implications if I do this. Thoughts @pefarrell ?